### ISSUE TITLE
Handle No Delivery flows

### ISSUE THREAD

**Issue by loudnate** [loudnate/openaps-mmhistorytools#36]
It's happened to me before where I've forgotten to act on the Fixed Prime screen and the pump remains "suspended". It would be good if this could be reflected in the reconciled/resolved history.

**Comment by loudnate** [comment]
Start time: earlier of No Delivery alarm or Rewind
End time: next prime or basal start (basal start events might be x23 only)

**Comment by channemann** [comment]
Does basal start actually coincide with delivery of insulin? I believe you get a record for basal rate start if there is a scheduled change even if it's not delivering. Perhaps you are talking about something else.

**Comment by loudnate** [comment]
Yeah if you check history after any restart in delivery there's a BasalProfileStart entry. Something makes me think this was only added for the x23 models though.

**Comment by loudnate** [comment]
Here's the flow for a No Delivery alarm:

```json
[
  {
    "_type": "Prime", 
    "type": "fixed", 
    "_description": "Prime 2015-10-08T18:23:39 head[5], body[0] op[0x03]", 
    "timestamp": "2015-10-08T18:23:39", 
    "_body": "", 
    "fixed": 0.5, 
    "_head": "0300050005", 
    "amount": 0.5, 
    "_date": "a79712080f"
  }, 
  {
    "_type": "Prime", 
    "type": "manual", 
    "_description": "Prime 2015-10-08T18:22:51 head[5], body[0] op[0x03]", 
    "timestamp": "2015-10-08T18:22:51", 
    "_body": "", 
    "fixed": 0.0, 
    "_head": "030000003f", 
    "amount": 6.3, 
    "_date": "b39632080f"
  }, 
  {
    "_type": "Rewind", 
    "_description": "Rewind 2015-10-08T18:22:07 head[2], body[0] op[0x21]", 
    "timestamp": "2015-10-08T18:22:07", 
    "_body": "", 
    "_head": "2100", 
    "_date": "879612080f"
  }, 
  {
    "_type": "ClearAlarm", 
    "_description": "ClearAlarm 2015-10-08T18:22:04 head[2], body[0] op[0x0c]", 
    "timestamp": "2015-10-08T18:22:04", 
    "_body": "", 
    "_head": "0c04", 
    "_date": "849612080f"
  }, 
  {
    "_type": "NoDelivery", 
    "_description": "NoDelivery 2015-10-08T18:21:41 head[4], body[0] op[0x06]", 
    "timestamp": "2015-10-08T18:21:41", 
    "_body": "", 
    "_head": "06040c1e", 
    "_date": "a99552480f"
  },
]
```

**Comment by channemann** [comment]
Yes, looks the same as what I saw, except I had a basal profile start
between the rewind and the first prime.

On Thu, Oct 8, 2015 at 6:42 PM, Nathan Racklyeft <notifications@github.com>
wrote:

> Here's the flow for a No Delivery alarm:
>
> [
>   {
>     "_type": "Prime",
>     "type": "fixed",
>     "_description": "Prime 2015-10-08T18:23:39 head[5], body[0] op[0x03]",
>     "timestamp": "2015-10-08T18:23:39",
>     "_body": "",
>     "fixed": 0.5,
>     "_head": "0300050005",
>     "amount": 0.5,
>     "_date": "a79712080f"
>   },
>   {
>     "_type": "Prime",
>     "type": "manual",
>     "_description": "Prime 2015-10-08T18:22:51 head[5], body[0] op[0x03]",
>     "timestamp": "2015-10-08T18:22:51",
>     "_body": "",
>     "fixed": 0.0,
>     "_head": "030000003f",
>     "amount": 6.3,
>     "_date": "b39632080f"
>   },
>   {
>     "_type": "Rewind",
>     "_description": "Rewind 2015-10-08T18:22:07 head[2], body[0] op[0x21]",
>     "timestamp": "2015-10-08T18:22:07",
>     "_body": "",
>     "_head": "2100",
>     "_date": "879612080f"
>   },
>   {
>     "_type": "ClearAlarm",
>     "_description": "ClearAlarm 2015-10-08T18:22:04 head[2], body[0] op[0x0c]",
>     "timestamp": "2015-10-08T18:22:04",
>     "_body": "",
>     "_head": "0c04",
>     "_date": "849612080f"
>   },
>   {
>     "_type": "NoDelivery",
>     "_description": "NoDelivery 2015-10-08T18:21:41 head[4], body[0] op[0x06]",
>     "timestamp": "2015-10-08T18:21:41",
>     "_body": "",
>     "_head": "06040c1e",
>     "_date": "a99552480f"
>   },
> ]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/loudnate/openaps-mmhistorytools/issues/36#issuecomment-146733760>
> .
>



-- 
Chris Hannemann
chris.hannemann@gmail.com

**Comment by loudnate** [comment]
```json
[
  {
    "_type": "Prime", 
    "_description": "Prime 2015-12-26T17:58:16 head[5], body[0] op[0x03]", 
    "timestamp": "2015-12-26T17:58:16", 
    "_body": "", 
    "_head": "0300050005", 
    "amount": 0.5, 
    "fixed": 0.5, 
    "type": "fixed", 
    "_date": "d03a111a0f"
  }, 
  {
    "_type": "Prime", 
    "_description": "Prime 2015-12-26T17:57:08 head[5], body[0] op[0x03]", 
    "timestamp": "2015-12-26T17:57:08", 
    "_body": "", 
    "_head": "030000000d", 
    "amount": 1.3, 
    "fixed": 0.0, 
    "type": "manual", 
    "_date": "c839311a0f"
  }, 
  {
    "_type": "Rewind", 
    "_description": "Rewind 2015-12-26T17:56:25 head[2], body[0] op[0x21]", 
    "timestamp": "2015-12-26T17:56:25", 
    "_body": "", 
    "_head": "2100", 
    "_date": "d938111a0f"
  },
]
```

Here's a case without the NoDelivery alarm. From my experience, basal insulin delivery resumes either at the point after the fixed prime or when the prime screen is cleared. I think it's the X23 models only that prevent basal from resuming until you clear the prime screen.
